### ISSUE TITLE
Create & sync stack lineage on proposals

### ISSUE THREAD

**Issue by tranhl** [git-town/git-town#3003]
## Problem
My team have bought into Git Town as our stacked workflow tool of choice. We're loving the workflow so far and the productivity gains from not being blocked. It's also drastically reduced PR size as contributors no longer feel the need to submit work in large chunks out of fear of being blocked. Huge wins overall 🎉 

However our one and only pain point is that we now have 30+ proposals open at any given time 😅. This makes it pretty hard to track which proposals belong to which stack. We like to review from the base up for any given stack, so this is a big point of friction for us during code review.

Our current method is to figure out what the base PR to any stack is (it'll be the one targeting the main branch), review it, take note of it's branch name, then find the PR that targets that branch. Rinse repeat until we're done. Refined Github makes this a little easier, but it's still not great:

![image](https://github.com/git-town/git-town/assets/24240419/c40ab69b-67ab-49cc-8dc0-448bde98ae27)

## Solution?

[graphite.dev](https://graphite.dev/) has a neat feature where it maintains a stack lineage for each proposal it creates in a stack.

![image](https://github.com/git-town/git-town/assets/24240419/d60e66e6-dc58-434e-a994-8b5e1a71d4be)

This allows your team to easily visualise & navigate through your stack on Github.

It would be super helpful if Git Town could provide a similar feature, perhaps built into `git-town sync` and `git-town propose`?

**Comment by tranhl** [comment]
Thanks for getting back! Happy to answer your questions.

Let's use the stack from the example image above as reference:

> Do you review all PRs in a stack together, or only the top-level PR in a stack?

We start from the base (`[Gameboard] Initializing an empty gameboard` in this case). We'd then work our way up the stack. So next would be `[Gameboard] Add moving logic to gameboard`, and so on.

Ideally we review as often as PRs are being created, so that stacks don't blow up in size and cause merge conflict grief for everyone.

> Do you ship all PRs in a stack together, or do you always only the top-level PR?

The author would then ship the entire stack one PR at a time, starting from the base. This means `[Gameboard] Initializing an empty gameboard` would be shipped first.

> The stack you share in your screenshot looks like it's an ongoing "workstream" with more PRs to come soon. Because all you have right now is an empty gameboard with a controller and some styling. I imaging subsequent PRs will add additional game elements like rules, maps, leaderboards, etc.

That's correct. However for us, not everything has to be a stack. We only encourage stacks if not creating one would otherwise block your work. It's better instead to work directly off the main branch if possible. 

So long story short, we review as much as we can (ideally all at once), and ship one by one.

> I also like the idea of attaching a comment that describes the position of this PR in the lineage. Git Town already displays this information on the CLI with the [switch command](https://www.git-town.com/commands/switch). It would make perfect sense to also include this information in PRs. Thanks for the suggestion! 🙏

Absolutely, what we're looking for is essentially a printout of `git-town switch` in each PR that has links to other PRs in the stack!

**Comment by tranhl** [comment]
FYI @kevgo we've created an internal GitHub action that generates the lineage & the PR description for each change in the stack with a printout as a workaround. This was achieved using the GitHub REST API.

The way it works is to have a pre-defined section in our pull request template for the stack lineage:

```md
## Stack

<autogenerated>

## ....
```

Then, everytime the PR changes (e.g. after running `git-town sync`), we get the PR's description using the [get a pull request API](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#get-a-pull-request). Using a regex, we could then regenerate the stack lineage and update the PR body, and then update the PR itself using the [update a pull request API](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#update-a-pull-request):

```ts
/**
 * Matches the `## Stack` markdown header and all of it's content until
 * the next markdown header or end of file.
 * */
const STACK_SECTION_REGEX = /^## Stack\s*[\s\S]*?(?=\n##|\n$)/m

let {
  data: { body: description },
} = await octokit.rest.pulls.get({
  ...github.context.repo,
  pull_number: currentPullRequest.number,
})

description = description ?? ''
description = description.replace(
  STACK_SECTION_REGEX,
  ['## Stack', '', output].join('\n')
)

await octokit.rest.pulls.update({
  ...github.context.repo,
  pull_number: currentPullRequest.number,
  body: description,
})
```

I believe a similar approach could be used by Git Town when pushing changes to branches during `git-town sync`. All that's necessary would be to allow users to configure the regex that Git Town would use to find-and-replace the stack lineage each sync. Since we now have configuration files, this would even work in a team setting.

Just some food for thought! 😊

**Comment by kevgo** [comment]
This is pretty brilliant, thanks for sharing! I especially like the idea to add the stack into the original PR description. This is MUCH better than describing the stack in a separate comment.

The advantage of your approach to use a GitHub Action over doing this with Git Town is that your implementation also works when somebody changes the base branch of a PR using the GitHub UI. Or uses any other tool. So I think using a GitHub action is the best way to implement this feature. If you agree, and your solution works well enough for you, feel free to close this issue.

Have you considered open-sourcing your action and publishing it in the [GitHub marketplace](https://github.com/marketplace?type=actions) so that others can use it as well?

**Comment by tranhl** [comment]
> Have you considered open-sourcing your action and publishing it in the [GitHub marketplace](https://github.com/marketplace?type=actions) so that others can use it as well?

This is definitely something that would benefit the community at large!
Is there any chance we could bring this under the `git-town` organisation as an official integration? People would be more likely to adopt the action if it was officially supported as opposed to being 3rd party.

**Comment by kevgo** [comment]
We can certainly host this under the `git-town` organization on GitHub! It would have to be its own repo because that's how the GitHub marketplace works. Do you have bandwidth to provide bug fixes and respond to tickets for it in the foreseeable future? I don't expect it to require much time since the action is a rather small codebase. The benefit for you would be that you can call yourself an official maintainer of Git Town!

**Comment by tranhl** [comment]
Awesome! I think I'll be able to dedicate bandwidth for the first month or two to help iron out any issues we encounter after rolling out the action. Afterwards it would be nice to take a back seat and do mostly triaging, as my schedule varies quite a bit, making it hard for me to commit to anything long term. How does that sound?

**Comment by kevgo** [comment]
Okay great! I have set up https://github.com/git-town/action and made you an admin there. Feel free to set this up any way you like and please ping me if you need anything!

**Comment by oleg-codaio** [comment]
Just my 2¢ but I'd still love a version of this that runs client side when pushing up PRs simply because it's a more accessible solution, especially in larger companies where setting up a GitHub Action would require broader buy-in from the team.

**Comment by ruudk** [comment]
The Action looks great, but it's a bit sad that it requires an action to work. I'd rather have a cli command that can generate this list that can I bind to an alias.

**Comment by kevgo** [comment]
Thanks for the feedback. Reopening this ticket for adding this feature to the CLI app.

**Comment by oleg-codaio** [comment]
BTW, if anyone needs inspiration feel free to reference this PR: https://github.com/aviator-co/av/pull/271

**Comment by Ydot19** [comment]
@kevgo Can I claim this issue if anyone else hasn't started to work on it?

**Comment by kevgo** [comment]
Go for it!

**Comment by tranhl** [comment]
@kevgo I recently landed https://github.com/git-town/action/pull/65 allowing stack visualizations in PR comments.

When displaying in the description, we're able to piggy-back off `pull_request_template.md` to allow customization of the visualization. Unfortunately there isn't an equivalent for comments so the visualization is bare bones right now:

<img width="936" height="186" alt="Image" src="https://github.com/user-attachments/assets/1a109b9e-0386-4100-9e54-885e2d10c414" />

I can see people wanting to customize the comment as well, however the approach we take would need to work for both the GitHub action & CLI.

## Solutions

1. We could introduce additional Git Town configuration to support this, either via config file or some other mechanism. I'm personally not a fan of this, as templates are usually multiline text making configuration via CLI clunky.
2. We could read the comment template from a standardized file (i.e. `.github/stack_comment_template.md`) that is functionally equivalent to `pull_request_template.md`. The GitHub action & CLI could read from this file and then render out the final stack lineage from there. This could be isolated to just the GitHub forge connector to avoid bloating the Git Town core.

This will impact #4871 so I'm keen to get your 2c on this!

**Comment by yaadata** [comment]
This feature is ready for people that want to pilot it/kick the wheels https://github.com/git-town/git-town/pull/4871. 

Per discussion, the pull-request updates the body of the pull-request (see an example in the pr). If we want pull-request comments to display the stack lineage, I did this in an earlier iteration and that can be a follow up to the PR mentioned above

**Comment by kevgo** [comment]
Release plan for this: https://github.com/git-town/git-town/discussions/5602

**Comment by JaredSharplin** [comment]
Loving this feature, thanks for everyone's hard work on it.

Just had a minor request - would it be possible to configure it so that stacks that are only one level deep aren't added to the PR? I find there's not much value in this case:

<img width="775" height="119" alt="Image" src="https://github.com/user-attachments/assets/b780ebb0-2258-4569-8cde-1d7101399408" />

**Comment by kevgo** [comment]
@JaredSharplin I like this idea. I also find single-branch stacks more distracting than helpful. If a branch is going straight into the main branch, that relationship is already clearly expressed by the PR's target branch. Repeating it via lineage doesn't add much value.

The Git Town GitHub Action calls this setting [skip-single-stacks](https://github.com/git-town/action?tab=readme-ov-file#skip-single-stacks). This makes sense in that context since the action’s sole responsibility is attaching lineage to proposals. For the Git Town CLI, though, the naming probably needs to be a bit more general. We don't want to skip single stacks altogether, just not attach lineage to their proposals.

```toml
[propose]
lineage = "cli"                 # make the Git Town executable add lineage
lineage-single-stacks = false   # don't add lineage for single-branch stacks
```

```sh
git config git-town.proposals-show-lineage-single-stacks false
```

**Comment by kevgo** [comment]
This is available in [v22.5](https://github.com/git-town/git-town/releases/tag/v22.5.0).
